starlette/_utils.py

```diff
@@ -10,3 +17,50 @@ def is_async_callable(obj: typing.Any) -> bool:
     return asyncio.iscoroutinefunction(obj) or (
         callable(obj) and asyncio.iscoroutinefunction(obj.__call__)
     )
+
+
+T_co = typing.TypeVar("T_co", covariant=True)
+
+
+class AwaitableOrContextManager(Protocol[T_co]):
+    def __await__(self) -> typing.Generator[typing.Any, None, T_co]:
+        ...  # pragma: no cover
+
+    async def __aenter__(self) -> T_co:
+        ...  # pragma: no cover
+
+    async def __aexit__(
+        self,
+        __exc_type: typing.Optional[typing.Type[BaseException]],
+        __exc_value: typing.Optional[BaseException],
+        __traceback: typing.Optional[TracebackType],
+    ) -> typing.Union[bool, None]:
+        ...  # pragma: no cover
```

Comment:
Because of https://github.com/python/typing_extensions/issues/34. This is unnecessary on 3.8+

Reply:
What part exactly? I didn't get it.

Reply:
Ah, got it. Should we add a note or a conditional so we can remove when 3.8 EOL happens?

Reply:
Nah let’s just remove the boilerplate when 3.8 becomes EOL; I don’t think it will be a breaking change at runtime

Reply:
I mean, I think we need a way for us to remember to remove by then :thinking:

Reply:
I don't think there's an easy way to remind ourselves of this.

Reply:
I added a comment. I think that's the best we can do.

Reply:
Yeah, that's what I meant. :+1: